[CK_TILE] Stream-K Gemm Example for fp8 and bf8

arai713 opened the pull request:
## Proposed changes
This PR is expanding the supported datatypes for Stream-K Gemm by adding examples for fp8 and bf8. Previously examples for fp16 and bf 16 were added through this [PR](https://github.com/ROCm/composable_kernel/pull/2864). Currently these examples only support atomic reduction. Unit tests for fp8 and bf8 have been added through a test suite.

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [x] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [ ] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [ ] I have added inline documentation which enables the maintainers with understanding the motivation
- [ ] I have removed the stale documentation which is no longer relevant after this pull request
- [ ] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [x] I have run `clang-format` on all changed files
- [x] Any dependent changes have been merged

arai713 committed:
Addition of streamk fp8 example for CK Tile

arai713 committed:
Adding in bf8 streamk example in CK Tile

ecamartins commented on example/ck_tile/CMakeLists.txt:
```diff
@@ -27,4 +27,4 @@ add_subdirectory(36_pooling)
 add_subdirectory(39_copy)
 add_subdirectory(40_streamk_gemm)
-add_subdirectory(41_batched_contraction)
+add_subdirectory(41_batched_contraction)
```
I think you accidently removed the empty line from this file. Can you please add it back?

ecamartins commented on test/ck_tile/gemm_streamk/test_gemm_streamk_types.hpp:
Do we want to overwrite the existing types, or just add a new type for BF8 and FP8? (Right now they are overwritten)

arai713 replied:
This has been addressed in [92f2fcc](https://github.com/ROCm/composable_kernel/pull/3041/commits/92f2fcc38acc1b92a03f03a2b96f56a4bc7635be).

ecamartins commented on test/ck_tile/gemm_streamk/test_gemm_streamk_types.hpp:
```diff
@@ -1,123 +1,29 @@
+    std::tuple<    Row,     Col,     Row,       F16,       F16,         F32,       F16>,
+    std::tuple<    Row,     Col,     Row,       F8,       F8,         F32,       F16>,
+    std::tuple<    Row,     Col,     Row,       BF8,       BF8,         F32,       F16>,
+    std::tuple<    Row,     Col,     Row,       BF16,      BF16,        F32,       BF16>
```
Assuming we don't want to overwrite the existing smoke tests, I think we can remove FP16 and BF16 from these types. Also, can we please add RRR, CRR, and CCR for FP8 and BF8?

arai713 replied:
These layouts are being tested in [92f2fcc](https://github.com/ROCm/composable_kernel/pull/3041/commits/92f2fcc38acc1b92a03f03a2b96f56a4bc7635be)

ecamartins commented on test/ck_tile/gemm_streamk/test_gemm_streamk_types_fp8_bf8.hpp:
```diff
@@ -0,0 +1,27 @@
+using Col = ck_tile::tensor_layout::gemm::ColumnMajor;
+
+// clang-format off
+using KernelTypesStreamK = ::testing::Types<
```
I think there is a duplicate of `KernelTypesStreamK` in test/ck_tile/gemm_streamk/test_gemm_streamk_types.hpp.

arai713 replied:
This file has been removed when refactoring the unit tests.

ecamartins commented on test/ck_tile/gemm_streamk/test_gemm_streamk_util_fp8_bf8.hpp:
```diff
@@ -0,0 +1,282 @@
+        const float max_accumulated_value =
+            *std::max_element(c_m_n_host_ref.mData.begin(), c_m_n_host_ref.mData.end());
+        const auto rtol_atol = calculate_rtol_atol<ADataType, BDataType, AccDataType, CDataType>(
+            K, /*kbatch*/ 1, max_accumulated_value);
```
We may need to use `estimate_num_wgs_per_tile` here because for some test cases, we have >1 workgroup atomically adding to the same C macro tile. This could lead to round-off error and result in failing tests in a sporadic nature.

ecamartins replied:
The refactor in [ffe459a](https://github.com/ROCm/composable_kernel/pull/3041/commits/ffe459a49dd96cbaa4fe33d24f9d5d49533df6b4) uses example/ck_tile/40_streamk_gemm/run_gemm_example.inc which uses `estimate_num_wgs_per_tile`, so no need to add this anymore.

ecamartins commented on example/ck_tile/40_streamk_gemm/gemm_utils.hpp:
```diff
@@ -75,6 +73,18 @@ struct DataTypeTraits<ck_tile::bf16_t>
+    static constexpr const char* name = "bf8";
+};
+
 auto create_args(int argc, char* argv[])
```
In this function and the README, can we please update our datatype support?
In this function, I'm referring to this line:
`.insert("prec", "fp16", "data type. fp16/bf16")`

We should update it to `.insert("prec", "fp16", "data type. fp16/bf16/fp8/bf8")`

ecamartins commented on example/ck_tile/40_streamk_gemm/streamk_gemm.cpp:
Since we already have an example that supports fp16 and bf16, doesn't think functionality already exist in example/ck_tile/40_streamk_gemm/streamk_gemm_basic.cpp?
If so, it is possible to remove this file?

cgmillette commented on test/ck_tile/gemm_streamk/test_gemm_streamk_util_fp8_bf8.hpp:
```diff
@@ -0,0 +1,282 @@
+// Copyright © Advanced Micro Devices, Inc., or its affiliates.
```
There is a test harness already in test_gemm_streamk.hpp, I don't think we need to duplicate it here

arai713 replied:
The refactored tests from [92f2fcc](https://github.com/ROCm/composable_kernel/pull/3041/commits/92f2fcc38acc1b92a03f03a2b96f56a4bc7635be) are using the test harness.

cgmillette commented on test/ck_tile/gemm_streamk/test_gemm_streamk_types.hpp:
```diff
@@ -1,123 +1,29 @@
 using Row = ck_tile::tensor_layout::gemm::RowMajor;
 using Col = ck_tile::tensor_layout::gemm::ColumnMajor;
 
-using Mem    = ck_tile::integral_constant<GemmPipelineType, GemmPipelineType::Mem>;
```
These are types needed for the rest of the smoke tests

arai713 replied:
This file has been restored in [92f2fcc](https://github.com/ROCm/composable_kernel/pull/3041/commits/92f2fcc38acc1b92a03f03a2b96f56a4bc7635be)

cgmillette commented on test/ck_tile/gemm_streamk/CMakeLists.txt:
```diff
@@ -3,6 +3,7 @@ if(GPU_TARGETS MATCHES "gfx9")
 
     include_directories(BEFORE ${CMAKE_CURRENT_SOURCE_DIR})
  
+    add_gtest_executable(test_ck_tile_streamk test_gemm_streamk_fp8_bf8.cpp)
```
We should look at adding the fp8 tests to the smoke test

arai713 replied:
fp8/bf8 have been added to the refactored unit tests from [92f2fcc](https://github.com/ROCm/composable_kernel/pull/3041/commits/92f2fcc38acc1b92a03f03a2b96f56a4bc7635be)

cgmillette reviewed (changes requested):
Need to consolidate new fp8 tests into smoke tests with existing test harness.

arai713 committed:
Refactoring fp8/bf8 unit tests

Refactored the unit tests for fp8/bf8 to utilize the test harness.
Implemented smoke tests with layouts: CCR, CRR, RCR, RRR for fp8/bf8.
The tests are using 128x128x32 for the tile configuration, as other
configurations revealed implementation gaps that are currently being
documented.

ecamartins reviewed (dismissed):
LGTM! Great job on refactoring the example and tests :)

cgmillette reviewed (dismissed):
This looks great! Approved pending CI!

arai713 commented on test/ck_tile/gemm_streamk/test_gemm_streamk_types_fp8.hpp:
```diff
@@ -0,0 +1,76 @@
+//  - DATA TYPE: F8
+//  - LAYOUT: RRR, RRC, RCR, RCC, CRR, CRC, CCR, CCC
+//  - PIPELINE_TYPE: Mem, CompV3, CompV4
+//  - Macro Tile: 128x128x32, 128x128x64, 128x128x128, 256x128x32, 256x128x64, 128x256x32, 128x256x64, 256x256x32, 256x256x64
```
Updated option in DECLARE_F8_PARAMS_ALL_LAYOUTS_ALL_SIZES

arai713 commented on example/ck_tile/40_streamk_gemm/gemm_utils.hpp:
```diff
@@ -41,8 +41,6 @@ struct GemmConfigMemoryInterwave : public GemmConfigBase
     static constexpr ck_tile::index_t M_Warp_Tile = 32;
     static constexpr ck_tile::index_t N_Warp_Tile = 32;
     static constexpr ck_tile::index_t K_Warp_Tile = sizeof(PrecType) == 2 ? 8 : 16;
```
Added the Scheduler back in to the GemmConfig

cgmillette reviewed (approved):
lgtm!
